presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayBlock.java

```diff
@@ -0,0 +1,258 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+
+public abstract class AbstractArrayBlock
+        implements Block
+{
+    static int checkedCast(long value)
+    {
+        int result = (int) value;
+        if (result != value) {
+            throw new IllegalArgumentException("Size is greater than maximum int value");
+        }
+        return result;
+    }
+
+    protected abstract Block getValues();
+    protected abstract Block getOffsets();
```

Comment:
Use a `Slice` instead of a `Block` since you don't need nullable positions, and a `Slice` is lighter weight